FabricObserver/Observers/AppObserver.cs

```diff
@@ -2455,11 +2689,120 @@ private async Task SetDeployedApplicationReplicaOrInstanceListAsync(Uri applicat
+                return (false, 0);
+            }
+
+            // Don't waste cycles with XML parsing if you can easily get a hint first..
+            if (!appManifestXml.Contains("<ResourceGovernancePolicy "))
+            {
+                return (false, 0);
+            }
+
+            // Safe XML pattern - *Do not use LoadXml*.
+            var appManifestXdoc = new XmlDocument { XmlResolver = null };
+
+            using (var sreader = new StringReader(appManifestXml))
+            {
+                using (var xreader = XmlReader.Create(sreader, new XmlReaderSettings { XmlResolver = null }))
+                {
+                    appManifestXdoc?.Load(xreader);
+
+                    ObserverLogger.LogInfo("Completed TupleGetResourceGovernanceInfo.");
+                    return TupleGetResourceGovernanceInfoFromAppManifest(ref appManifestXdoc, servicePkgName);
+                }
+            }
+        }
+
+        private (bool IsRGEnabled, double MemoryLimitMb) TupleGetResourceGovernanceInfoFromAppManifest(ref XmlDocument xDoc, string servicePkgName)
+        {
+            ObserverLogger.LogInfo("Starting TupleGetResourceGovernanceInfoFromAppManifest.");
+           
+            if (xDoc == null)
+            {
+                ObserverLogger.LogInfo("Completed TupleGetResourceGovernanceInfoFromAppManifest: xDoc == null.");
+                return (false, 0);
+            }
+
+            try
+            {
+                // Ensure you get the correct manifest import for specified service package name. There will generally be multiple RG limits set per application (so, per service settings).
+                var sNode = 
+                    xDoc.DocumentElement?.SelectSingleNode(
+                        $"//*[local-name()='ServiceManifestImport'][*[local-name()='ServiceManifestRef' and @*[local-name()='ServiceManifestName' and . ='{servicePkgName}']]]");
+                
+                if (sNode == null)
+                {
+                    ObserverLogger.LogInfo($"Completing TupleGetResourceGovernanceInfoFromAppManifest: Missing ServiceManifestImport for {servicePkgName}.");
+                    return (false, 0);
+                }
+
+                XmlNodeList childNodes = sNode.ChildNodes;
+                XmlNode rgNode = null;
+
+                foreach (XmlNode node in childNodes)
+                {
+                    if (node.Name == "Policies")
+                    {
+                        foreach (XmlNode polNode in node.ChildNodes)
+                        {
+                            if (polNode.Name == "ResourceGovernancePolicy")
+                            {
+                                rgNode = polNode;
+                                break;
+                            }
+                        }
+                        break;
+                    }
+                }
+
+                var memAttr = rgNode?.Attributes["MemoryInMBLimit"];
```

Comment:
Important: can we please also parse "MemoryInMB" parameter?

There is an option to use SF RG per code package in 3 different ways:
1. specifying MemoryInMB only - SF will assign provided value of MemoryInMB to the MemoryInMBLimit.
2. specifying MemoryInMBLimit only - SF will use provided value as the RG limit
3. specifying MemoryInMB and MemoryInMBLimit - SF will use provided value of MemoryInMBLimit as RG limit

For the first case we would need to parse MemoryInMB and to have logic at the FO side that will map the SF logic.

Edit: adding [link](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-resource-governance#specify-resource-governance)

Reply:
For sure. We like to have multiple ways to do the same thing! ... :) FO will handle those cases as well. Great catch.